### PR TITLE
[VBLOCKS-3227] ci extension e2e

### PR DESCRIPTION
**Pull request by kpchoy**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [VBLOCKS-3227](https://twilio-engineering.atlassian.net/browse/VBLOCKS-3227)

### Description

Implement extension e2e in chrome

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -103,6 +103,23 @@ commands:
...
+      - run:
+          name: "Extension: server start"
+          background: true
+          command: |
+            echo 'export APPLICATION_SID_EXTENSION="$APPLICATION_SID_EXTENSION"' >> "$BASH_ENV"
+            echo 'export ACCOUNT_SID="$ACCOUNT_SID"' >> "$BASH_ENV"
+            echo 'export API_KEY_SID="$API_KEY_SID"' >> "$BASH_ENV"
+            echo 'export API_KEY_SECRET="$API_KEY_SECRET"' >> "$BASH_ENV"
+            source "$BASH_ENV"
+            npm run start:extension:server
```
Comment (charliesantos) [inline .circleci/config.yml:117]: Can we put this in a script file and call the script here instead? See how we're doing relay server. Also, once it's in a script, do we still need these environment variables to be exported?

File: tests/extension/app/offscreen/offscreen.js [hunk 2]
```diff
@@ -33,6 +32,10 @@ async function setupDevice(identity, recepient, connectToken) {
   // when making an outgoing call. If it exists, we initiate the call right away.
   if (recepient) { 
     call = await device.connect({ params: { recepient } });
+    // [e2e-testing]: Log errors to terminal
+    call.on('error', (error) => {
+      chrome.runtime.sendMessage({type: 'offscreen', message: `offscreen.js error: ${JSON.stringify(error)}`})
+    });
```
Comment (charliesantos) [inline tests/extension/app/offscreen/offscreen.js:38]: Will the test fail if there's an error? I think we should fail it.

File: tests/extension/app/offscreen/offscreen.js [hunk 3]
```diff
@@ -33,6 +32,10 @@ async function setupDevice(identity, recepient, connectToken) {
   // when making an outgoing call. If it exists, we initiate the call right away.
   if (recepient) { 
     call = await device.connect({ params: { recepient } });
+    // [e2e-testing]: Log errors to terminal
+    call.on('error', (error) => {
+      chrome.runtime.sendMessage({type: 'offscreen', message: `offscreen.js error: ${JSON.stringify(error)}`})
```
Comment (mhuynh5757) [inline tests/extension/app/offscreen/offscreen.js:37]:
```suggestion
      chrome.runtime.sendMessage({ type: 'offscreen', message: `offscreen.js error: ${JSON.stringify(error)}` });
```

File: tests/extension/app/popup/popup.js [hunk 4]
```diff
@@ -26,6 +26,19 @@ function start() {
   chrome.runtime.sendMessage({ type: 'getstate' }, ({ status }) => {
     setStatus(status);
   });
+  
+  // [e2e-testing]: Since puppeteer only monitors logs for popup.js, we need
+  // to listen for error messages from other js files
+  chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
+    if (request.type === 'offscreen') {
+      console.log(request.message)
```
Comment (mhuynh5757) [inline tests/extension/app/popup/popup.js:34]:
```suggestion
      console.log(request.message);
```

File: tests/extension/app/popup/popup.js [hunk 5]
```diff
@@ -26,6 +26,19 @@ function start() {
...
+    }
+  })
```
Comment (mhuynh5757) [inline tests/extension/app/popup/popup.js:36]:
```suggestion
  });
```

File: tests/extension/app/popup/popup.js [hunk 6]
```diff
@@ -26,6 +26,19 @@ function start() {
...
+  chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
+    if (request.type === 'worker') {
+      console.log(request.message)
```
Comment (mhuynh5757) [inline tests/extension/app/popup/popup.js:39]:
```suggestion
      console.log(request.message);
```

File: tests/extension/app/popup/popup.js [hunk 7]
```diff
@@ -26,6 +26,19 @@ function start() {
...
+    }
+  })
```
Comment (mhuynh5757) [inline tests/extension/app/popup/popup.js:41]:
```suggestion
  });
```

File: tests/extension/app/popup/popup.js [hunk 8]
```diff
@@ -26,6 +26,19 @@ function start() {
...
   });
+  
+  // [e2e-testing]: Since puppeteer only monitors logs for popup.js, we need
+  // to listen for error messages from other js files
+  chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
+    if (request.type === 'offscreen') {
+      console.log(request.message)
+    }
+  })
+  chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
```
Comment (mhuynh5757) [inline tests/extension/app/popup/popup.js:37]: Is there any reason we don't have a single listener and a conditional `if (request.type === 'offscreen' || request.type === 'worker') {`?

File: tests/extension/tests/index.js [hunk 9]
```diff
@@ -74,13 +77,13 @@ describe('Chrome extension tests', function () {
 
   it('should allow worker.js to make outgoing call, and receive incoming call', async () => {
     const initButton = await page.$('#init');
+    await delay(1000); // allow time for page to render
```
Comment (charliesantos) [inline tests/extension/tests/index.js:80]: Is there an event that we can listen to when the page finished rendering? Adding an arbitrary delay may cause flakiness.

File: tests/extension/app/worker.js [hunk 10]
```diff
@@ -61,6 +61,9 @@ async function init() {
    */
 
   device.on('incoming', (call) => {
+    // [e2e-testing]: Log call info
+    chrome.runtime.sendMessage({type: 'worker', message: `worker.js: ${JSON.stringify(call.parameters)}`})
```
Comment (mhuynh5757) [inline tests/extension/app/worker.js:65]:
```suggestion
    chrome.runtime.sendMessage({ type: 'worker', message: `worker.js: ${JSON.stringify(call.parameters)}` });
```

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 11]
```diff
@@ -103,6 +103,23 @@ commands:
...
+    steps:
+      - run:
+          name: "Extension: server start"
+          background: true
+          command: |
+            echo 'export APPLICATION_SID_EXTENSION="$APPLICATION_SID_EXTENSION"' >> "$BASH_ENV"
+            echo 'export ACCOUNT_SID="$ACCOUNT_SID"' >> "$BASH_ENV"
+            echo 'export API_KEY_SID="$API_KEY_SID"' >> "$BASH_ENV"
+            echo 'export API_KEY_SECRET="$API_KEY_SECRET"' >> "$BASH_ENV"
+            source "$BASH_ENV"
```
Comment (mhuynh5757) [inline .circleci/config.yml:116]: These lines should be unnecessary. We're basically reading env vars, and then setting env vars. So we're redeclaring the same env vars with the same values and these lines are not really doing anything.

File: tests/extension/tests/index.js (hunk 9 above)
Comment (kpchoy) [reply tests/extension/tests/index.js:80]: Technically it should be finished rendering by line 79. I saw it flake once but that might have been from previous config. Let me remove this delay for now.

File: .circleci/config.yml (hunk 1 above)
Comment (kpchoy) [reply .circleci/config.yml]: Forgot to clean this up, export not necessary

File: .circleci/config.yml [hunk 12]
```diff
@@ -103,6 +103,23 @@ commands:
...
+            npm run start:extension:server
+  extension-puppeteer-test:
+    steps:
+      - run:
+          name: "Extension: Puppeteer test"
+          command: npm run test:extension
```
[the remark this answers is not in the document]
Comment (kpchoy) [reply .circleci/config.yml]: One section looks clean 👍

File: tests/extension/app/popup/popup.js (hunk 8 above)
Comment (kpchoy) [reply tests/extension/app/popup/popup.js:32]: Good point, but removed this listener in favor of just throwing error in `offscreen.js`.

File: .circleci/config.yml (hunk 11 above)
Comment (kpchoy) [reply .circleci/config.yml]: Good catch, export not necessary 👍

File: tests/extension/app/offscreen/offscreen.js [hunk 13]
```diff
@@ -26,15 +26,15 @@ async function setupDevice(identity, recepient, connectToken) {
   /**
    * NOTE(kchoy): This is the second device created in this extension. device2ClientIdentity
    * calls device1ClientIdentity in worker.js
-   */ 
+   */
 
   // The recepient parameter is provided to the offscreen document
   // when making an outgoing call. If it exists, we initiate the call right away.
   if (recepient) {
```
Comment (mhuynh5757) [inline tests/extension/app/offscreen/offscreen.js:33]:
```suggestion
  if (recipient) {
```
Nitpick

**Review by mhuynh5757 — APPROVED** [review]

**Review by charliesantos — APPROVED** [review]
This is great!

**Pull request merged**